craft_parts/utils/process.py

```diff
@@ -0,0 +1,184 @@
+        self.read_fd = read_fd
+        self.write_fd = write_fd
+
+        self._linebuf = bytearray()
+        self._streambuf = b""
+
+    @property
+    def singular(self) -> bytes:
+        return self._streambuf
+
+    def process(self) -> bytes:
+        """Process any data in ``self.read_fd``, then return it."""
+        data = os.read(self.read_fd, _BUF_SIZE)
+        i = data.rfind(b"\n")
+        if i >= 0:
+            self._linebuf.extend(data[: i + 1])
+            self._streambuf += self._linebuf
+            os.write(self.write_fd, self._linebuf)
+            self._linebuf.clear()
+            self._linebuf.extend(data[i + 1 :])
+            return data
+        self._linebuf.extend(data)
+        return b""
+
+
+def run(
+    command: Command,
+    *,
+    cwd: Path | None = None,
+    stdout: Stream = None,
+    stderr: Stream = None,
+    check: bool = True,
+) -> ProcessResult:
+    """Execute a subprocess and collect its output.
+
+    This function collects the stdout and stderr streams as separate
+    accounts and a singular, combined account.
+
+    :param command: Command to execute.
+    :type Command:
+    :param cwd: Path to execute in.
+    :type Path | None:
+    :param stdout: Handle to a fd or I/O stream to treat as stdout. None defaults
+        to ``sys.stdout``, and process.DEVNULL can be passed for no printing.
+    :type Stream:
+    :param stderr: Handle to a fd or I/O stream to treat as stderr. None defaults
+        to ``sys.stderr``, and process.DEVNULL can be passed for no printing.
+    :type Stream:
+    :param check: If True, a ProcessError exception will be raised if ``command``
+        returns a non-zero return code.
+    :type bool:
+
+    :raises ProcessError: If process exits with a non-zero return code.
+    :raises OSError: If the specified executable is not found.
+
+    :return: A description of the process' outcome.
+    :rtype: ProcessResult
+    """
+    proc = subprocess.Popen(
+        command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=cwd
+    )
+
+    out_fd = _select_stream(stdout, sys.stdout.fileno())
+    err_fd = _select_stream(stderr, sys.stderr.fileno())
+
+    # stdout and stderr are guaranteed not `None` because we called with `subprocess.PIPE`
+    proc_stdout = cast(IO[bytes], proc.stdout).fileno()
+    proc_stderr = cast(IO[bytes], proc.stderr).fileno()
+
+    os.set_blocking(proc_stdout, False)
+    os.set_blocking(proc_stderr, False)
+
+    out_handler = _ProcessStream(proc_stdout, out_fd)
+    err_handler = _ProcessStream(proc_stderr, err_fd)
+    combined = b""
+
+    while True:
+        r, _, _ = select.select([proc_stdout, proc_stderr], [], [])
+
+        try:
+            if proc_stdout in r:
+                combined += out_handler.process()
+
+            if proc_stderr in r:
+                combined += err_handler.process()
+
+        except BlockingIOError:
+            pass
+
+        except Exception:
```

Comment:
Any other errors that would arise from this would be strange system-level errors such as pipes being dropped or files becoming inaccessible.

Are you wanting to try to recover from more errors, or do you want to attach more information to errors before raising them? I don't think there's any other errors that can be recovered from, but I can attach extra info if you have specific info in mind.

Reply:
I can see that you're doing that only to close `/dev/null` fd if that's the case.
Maybe `_select_stream` could return a context manager that would do nothing for the regular stream and `os.close()` for `DEVNULL` and then you could skip this codeblock entirely (and also the one after the while)